src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerExecutor.cs

```diff
@@ -1366,52 +997,26 @@ public bool TryMarkSymbolEndAnalysisComplete(ISymbol symbol, DiagnosticAnalyzer
             Func<SyntaxNode, TLanguageKindEnum> getKind,
             Action<Diagnostic> addDiagnostic,
             Func<Diagnostic, bool> isSupportedDiagnostic,
-            bool isGeneratedCode,
-            SyntaxNodeAnalyzerStateData? analyzerState)
+            bool isGeneratedCode)
             where TLanguageKindEnum : struct
         {
             Debug.Assert(nodeActionsByKind.Any());
             Debug.Assert(!isGeneratedCode || !_shouldSkipAnalysisOnGeneratedCode(analyzer));
             Debug.Assert(!IsAnalyzerSuppressedForTree(analyzer, model.SyntaxTree));
 
-            var partiallyProcessedNode = analyzerState?.CurrentNode;
-            if (partiallyProcessedNode != null)
-            {
-                ExecuteSyntaxNodeActions(partiallyProcessedNode, nodeActionsByKind, containingSymbol, model, getKind, addDiagnostic, isSupportedDiagnostic, isGeneratedCode, analyzerState);
-            }
-
-            foreach (var child in nodesToAnalyze)
-            {
-                if (ShouldExecuteNode(analyzerState, child, analyzer))
-                {
-                    SetCurrentNode(analyzerState, child);
-
-                    ExecuteSyntaxNodeActions(child, nodeActionsByKind, containingSymbol, model, getKind, addDiagnostic, isSupportedDiagnostic, isGeneratedCode, analyzerState);
-                }
-            }
-        }
-
-        private void ExecuteSyntaxNodeActions<TLanguageKindEnum>(
-            SyntaxNode node,
-            ImmutableSegmentedDictionary<TLanguageKindEnum, ImmutableArray<SyntaxNodeAnalyzerAction<TLanguageKindEnum>>> nodeActionsByKind,
-            ISymbol containingSymbol,
-            SemanticModel model,
-            Func<SyntaxNode, TLanguageKindEnum> getKind,
-            Action<Diagnostic> addDiagnostic,
-            Func<Diagnostic, bool> isSupportedDiagnostic,
-            bool isGeneratedCode,
-            SyntaxNodeAnalyzerStateData? analyzerState)
-            where TLanguageKindEnum : struct
-        {
-            if (nodeActionsByKind.TryGetValue(getKind(node), out var actionsForKind))
+            foreach (var node in nodesToAnalyze)
             {
-                foreach (var action in actionsForKind)
+                if (ShouldExecuteNode(node, analyzer))
                 {
-                    ExecuteSyntaxNodeAction(action, node, containingSymbol, model, addDiagnostic, isSupportedDiagnostic, isGeneratedCode, analyzerState);
+                    if (nodeActionsByKind.TryGetValue(getKind(node), out var actionsForKind))
```

Comment:
❔ It seems like it's more likely for `actionsForKind` to be empty than for `ShouldExecuteNode` to be false. Should we reorder these checks and avoid calling `ShouldExecuteNode` on syntax nodes that don't have any registered actions?

Reply:
➡️ Submitted as follow-up #66923